### ISSUE TITLE
cleanup supervised training pipeline

### ISSUE THREAD

**Issue by Talor-A** [Talor-A/forge#7]
I have enough to deliver the map. Here is the analysis.

---

# Supervised Training Pipeline — Map & Audit

The supervised pipeline goes **Java game engine → JSONL trajectories → Python preprocess → mmap arrays → PyTorch model → checkpoint → ONNX → Java inference**. RL/PPO/AWR/MCTS pieces branch off this trunk but aren't strictly part of the supervised path.

### 1. Headless game collection (Java)

| Component                        | File                                                                 | Lines           |
| -------------------------------- | -------------------------------------------------------------------- | --------------- |
| CLI entry point                  | `forge-gui-desktop/src/main/java/forge/view/SimulateRLTraining.java` | **956** ⚠️      |
| Parallel game runner (alt entry) | `forge-ai-rl/.../training/GameRunner.java`                           | 274             |
| Lobby + controller stubs         | `forge-ai-rl/.../LobbyPlayerRL.java`, `PlayerControllerRL.java`      | 46 / **801** ⚠️ |
| RL config                        | `forge-ai-rl/.../RLConfig.java`                                      | 85              |
| Central orchestrator             | `forge-ai-rl/.../RLController.java`                                  | **802** ⚠️      |

**Input:** CLI args (`-d deck1.dck -d deck2.dck -n 1000 -t 16 -o trajectories/ -q`), the Forge `res/` tree, decks in `~/.forge/decks/constructed/`.
**Output:** one `traj_*.jsonl` per player per game in `outputDir`.
**Mode used for supervised data:** `rltrain collect` → `RECORD_HEURISTIC` (PlayerControllerRL delegates every decision to the heuristic AI and records the result).

There are also `mcts-collect`, `evaluate`, `selfplay`, `leagueplay`, `headtohead` modes — orthogonal to supervised training but all share the same `runSingleGame()` plumbing.

### 2. Per-decision feature encoding (Java, called during step 1)

| Component                      | File                                                    | Lines      | Output dim                                                |
| ------------------------------ | ------------------------------------------------------- | ---------- | --------------------------------------------------------- |
| Global + zone encoder          | `features/GameStateEncoder.java`                        | 430        | 96 global + 145 cards × 256 = **37,216** flat floats      |
| Per-card features              | `features/CardFeatures.java`                            | **529** ⚠️ | 256                                                       |
| Action features (SpellAbility) | `features/ActionEncoder.java`                           | 275        | 64                                                        |
| Combat math features           | `features/CombatMath.java`                              | **499** ⚠️ | injected into card features [208–231] and globals [76–95] |
| Shape container                | `features/GameStateFeatures.java`                       | 116        | —                                                         |
| Normalizer (unused?)           | `features/FeatureNormalizer.java`                       | 42         | —                                                         |
| Decision enum                  | `decisions/DecisionType.java`                           | 37         | 10 types                                                  |
| Decision DTOs                  | `decisions/DecisionContext.java`, `DecisionResult.java` | 69 / 51    | —                                                         |

### 3. Trajectory writer (Java)

| Component                                                 | File                               | Lines      |
| --------------------------------------------------------- | ---------------------------------- | ---------- |
| JSONL writer                                              | `training/TrajectoryRecorder.java` | 187        |
| Listener that pulls features per decision                 | `TrajectoryDecisionListener.java`  | 192        |
| Direct GameEvent subscriber (mostly unused per CLAUDE.md) | `GameStateRecorder.java`           | 268        |
| Subclass recorder (in-tree, unclear if hot)               | `RecordingControllerAi.java`       | 431        |
| Wrapper recorder (older, large)                           | `RecordingPlayerController.java`   | **789** ⚠️ |
| Marker subclass                                           | `RecordingLobbyPlayerAi.java`      | 36         |
| Human-game capture                                        | `training/HumanGameRecorder.java`  | 375        |

**Output file format:**
```
{header: gameId, won, totalDecisions, durationMs}
{turnIndex, decisionType, contextInfo, globalFeatures[96],
 gameStateFlat[37216], candidateFeatures[N][256 or 64], candidateCount,
 selectedIndices[k], actionProbabilities[N], visitProportions[N],
 valueEstimate, usedFallback, intermediateReward, terminalReward,
 spellFeatures[64]}
```

⚠️ **Three recording paths coexist** (`RecordingPlayerController`, `RecordingControllerAi`, `TrajectoryDecisionListener` via `PlayerControllerRL`). CLAUDE.md and `SimulateRLTraining.runSingleGame()` indicate the listener via `PlayerControllerRL` is the live one — the other two are likely dead code candidates for deletion.

### 4. Preprocessing JSONL → memory-mapped numpy (Python)

| Component                    | File                                  | Lines      |
| ---------------------------- | ------------------------------------- | ---------- |
| Preprocessor                 | `training/preprocess_trajectories.py` | **669** ⚠️ |
| Collection dashboard wrapper | `training/collect_ui.py`              | 630        |

**Input:** dir of `traj_*.jsonl`.
**Output (laid out under `preprocessed/`):**
```
shared/{game_state.npy[N×37216], global_features.npy[N×96], outcome.npy[N], file_id.npy[N]}
priority/{gs_index, candidates[N×MAX×64], candidate_mask, selected_idx, action_probs}
attack/{gs_index, creatures[N×MAX×256], creature_mask, action_mask, action_probs}
block/{gs_index, pairs[N×MAX×512], pair_mask, action_mask, action_probs}
target/{gs_index, candidates[N×MAX×?], candidate_mask, selected_idx, spell_features[64]}
card_select/{...}  mulligan/{...}  binary/{...}
metadata.json
```

Game state is stored once per snapshot; each decision-type table indexes back via `gs_index`. GAE-style returns (γ=0.99) are computed here from `intermediateReward + terminalReward`.

### 5. Dataset loaders (Python) — **TWO PARALLEL PATHS, big red flag**

| Component                      | File                         | Lines | Used by                                                                  |
| ------------------------------ | ---------------------------- | ----- | ------------------------------------------------------------------------ |
| Memory-mapped path (canonical) | `training/mmap_dataset.py`   | 573   | `train_decisions_ui.py` (Mmap* classes)                                  |
| Raw-JSONL path (legacy)        | `training/chunked_loader.py` | 319   | `train_value.py`, `train_decisions.py`, parts of `train_decisions_ui.py` |

`parse_game_state()` exists in `mmap_dataset.py` and contains the boundary assertion. But `chunked_loader.load_value_samples` reimplements the same zone parsing inline (lines 145–162), and `train_decisions.py` reimplements it again — three copies of the same logic.

### 6. Model (Python)

| Component                                                  | File                                                                                                                                                                        | Lines   |
| ---------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------- |
| Top-level model                                            | `model/mtg_model.py`                                                                                                                                                        | 156     |
| Shared state encoder (set-attention per zone + cross-zone) | `model/game_state_encoder.py`                                                                                                                                               | 161     |
| Value head                                                 | `model/value_network.py`                                                                                                                                                    | 37      |
| Decision heads                                             | `priority_head.py` 94, `combat_attack_head.py` 98, `combat_block_head.py` 114, `target_head.py` 113, `card_select_head.py` 102, `mulligan_head.py` 115, `binary_head.py` 41 | small ✓ |
| Auto-detected GPU profile                                  | `model/gpu_config.py`                                                                                                                                                       | 164     |

Sizes: small (~11M) → xl (~107M). Input: `(global[96], my_board[40×256] + mask, opp_board[40×256] + mask, hand[15×256], my_gy[20×256], opp_gy[20×256], stack[10×256])`.

### 7. Training loops (Python)

| Component                             | File                             | Lines         |
| ------------------------------------- | -------------------------------- | ------------- |
| Value-network training                | `training/train_value.py`        | 549           |
| Decision-head training (headless)     | `training/train_decisions.py`    | **783** ⚠️    |
| Decision-head training + Tk dashboard | `training/train_decisions_ui.py` | **2730** ⚠️⚠️ |
| AWR trainer                           | `training/awr_trainer.py`        | 641           |
| PPO trainer (out of supervised path)  | `training/ppo_trainer.py`        | **1719** ⚠️   |
| PPO Tk UI                             | `training/ppo_ui.py`             | **1536** ⚠️   |
| Misc monitoring                       | `monitor_ui.py` 281              | —             |

Checkpoint: PyTorch `.pt` saving `{config, state_dict}`. `MTGModel.load` silently skips weights whose shapes don't match — a known footgun.

### 8. Export to ONNX (Python)

| Component     | File                   | Lines |
| ------------- | ---------------------- | ----- |
| ONNX exporter | `tools/export_onnx.py` | 376   |

Produces 9 `.onnx` files (+ `.data` weight files) → committed under `forge-ai-rl/models/`.

### 9. Serving / Java inference

| Component                              | File                           | Lines |
| -------------------------------------- | ------------------------------ | ----- |
| Python TCP/length-prefixed JSON server | `serving/model_server.py`      | 615   |
| Java client over TCP                   | `model/ModelServerClient.java` | 163   |
| Java ONNX runtime client               | `model/ONNXModelClient.java`   | 618   |

---

## Long-file watchlist (likely AI-written, prime bug surface)

| Rank | File                                  | Lines |
| ---- | ------------------------------------- | ----- |
| 1    | `training/visualize_game_state.py`    | 2806  |
| 2    | `training/train_decisions_ui.py`      | 2730  |
| 3    | `training/ppo_trainer.py`             | 1719  |
| 4    | `training/ppo_ui.py`                  | 1536  |
| 5    | `view/SimulateRLTraining.java`        | 956   |
| 6    | `mcts/MCTSDecisionMaker.java`         | 949   |
| 7    | `RLController.java`                   | 802   |
| 8    | `PlayerControllerRL.java`             | 801   |
| 9    | `RecordingPlayerController.java`      | 789   |
| 10   | `training/train_decisions.py`         | 783   |
| 11   | `training/preprocess_trajectories.py` | 669   |
| 12   | `training/awr_trainer.py`             | 641   |
| 13   | `model/ONNXModelClient.java`          | 618   |
| 14   | `serving/model_server.py`             | 615   |
| 15   | `mmap_dataset.py`                     | 573   |
| 16   | `train_value.py`                      | 549   |
| 17   | `features/CardFeatures.java`          | 529   |
| 18   | `features/CombatMath.java`            | 499   |
| 19   | `RecordingControllerAi.java`          | 431   |
| 20   | `features/GameStateEncoder.java`      | 430   |

`train_decisions_ui.py` at 2,730 lines mixing GUI, dataset selection (both loaders), and 7 training routines is almost certainly the worst test surface. `preprocess_trajectories.py` (669) writes the most consequential artifact in the system and has *zero* tests.

---

## Data contracts — how they're upheld

| Contract                                                               | Where declared                                                                                                                                     | How enforced                                                                                                                                   |
| ---------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------- |
| `globalFeatures` = 96 floats                                           | `GameStateEncoder.GLOBAL_FEATURE_SIZE`, `mmap_dataset.GLOBAL_DIM`                                                                                  | Asserted in `parse_game_state` and `_process_file` (lengths only ∈ {0, 96}). `chunked_loader.load_value_samples` silently truncates/zero-pads. |
| `gameStateFlat` = 37,216 floats (96 + 145·256)                         | `mmap_dataset.GAME_STATE_DIM` (canonical), recomputed in `train_decisions.py`, `chunked_loader.py`, `preprocess_trajectories.py`, `export_onnx.py` | Asserted in `parse_game_state` and `_process_file`. Other paths silently pad/truncate.                                                         |
| `candidateFeatures` row dim ∈ {64, 256, 512} depending on decisionType | Implicit (no central constant)                                                                                                                     | Magic numbers in `_scan_one` (`len(cand[0]) > CARD_DIM + 10` to distinguish block format) — duck-typing of file format.                        |
| 9 ONNX files match Java inference shapes                               | `export_onnx.py` MAX_* constants vs Java `ONNXModelClient`                                                                                         | No automated check; drift would be silent until evaluation.                                                                                    |
| Train/val split disjoint by game                                       | `_split_file_ids` in `mmap_dataset.py`, `split_by_game` in `chunked_loader.py`                                                                     | Two implementations; not tested.                                                                                                               |

The 37,216 dim is hard-coded in ≥5 places; renaming or growing the feature vector would require touching every site.

---

## Tests that exist (very thin)

| Test                                        | File                                                                | What it actually checks                                                                                         |
| ------------------------------------------- | ------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------- |
| Value-net overfit / grad-flow / determinism | `src/test/python/test_value_network_overfit.py` (97)                | The single best diagnostic per `RL_BEST_PRACTICES.md`; only covers `ValueNetwork` in isolation, not `MTGModel`. |
| GAE / discounted-return math                | `src/test/python/test_gae.py` (124)                                 | Hand-computed 3-step trajectory + geometric-decay edge cases.                                                   |
| MCTS rollout listener / graph               | `test_rollout_log.py` (203), `test_rollout_graph.py` (208)          | Not part of supervised pipeline.                                                                                |
| `canPlayForRL` classification               | `src/test/java/forge/ai/rl/CanPlayForRLTest.java` (135)             | Strategic-veto vs mechanical-failure enum classification, not end-to-end.                                       |
| MCTS Java tests                             | `JsonlRolloutListenerTest.java` (205), `MCTSTimeoutTest.java` (179) | MCTS-only.                                                                                                      |

**Zero tests cover:** any Java feature encoder, `TrajectoryRecorder` round-trip, `preprocess_trajectories.py`, any `Mmap*Dataset`, any non-value head, ONNX export round-trip, the model server protocol. The system relies entirely on `smoke.sh` for "doesn't crash" assurance.

`forge-ai-rl/TESTING_PLAN.md` already enumerates the missing tests by priority — it identifies the same gaps and ranks GAE / overfit-batch / feature golden tests / checkpoint round-trip / PPO smoke as the top five (the first two are now done).

---

## End-to-end scripts

| Script                                                                                                | Scope                                                                                        | Notes                                                                                                                                   |
| ----------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------- |
| `smoke.sh`                                                                                            | Build assumed; collect 10 self-play games → train value head 30 epochs → assert 5× loss drop | **The only true end-to-end check.** Bypasses `preprocess_trajectories.py` (`train_value.py` reads JSONL directly via `chunked_loader`). |
| `scripts/02_collect_data.sh` → `collect_ui.py` → `SimulateRLTraining collect`                         | Step 1: collect                                                                              | Hard-coded `/home/maustin` paths.                                                                                                       |
| `scripts/02b_preprocess_data.sh`                                                                      | Step 2: JSONL → mmap                                                                         | Hard-coded `/home/maustin` paths.                                                                                                       |
| `scripts/03_train_value.sh` → `train_value.py`                                                        | Step 3: value head                                                                           | Reads JSONL, **not** the mmap output of 02b.                                                                                            |
| `scripts/04_train_decisions.sh` → `train_decisions.py`                                                | Step 4: decision heads                                                                       | Reads JSONL, not mmap.                                                                                                                  |
| `scripts/05_eval.sh`, `05_verify_model.sh`                                                            | Eval / verify                                                                                |                                                                                                                                         |
| `scripts/08_export_onnx.sh` → `tools/export_onnx.py`                                                  | ONNX export                                                                                  | Hard-coded `/home/maustin`.                                                                                                             |
| `forge-ai-rl/scripts/{collect_data,train,train_decisions,train_decisions_ui,evaluate,serve_model}.sh` | Older per-step wrappers                                                                      | Compute paths relative to `$SCRIPT_DIR`.                                                                                                |
| `mcts-smoke-test.sh`                                                                                  | MCTS-specific                                                                                | Not supervised path.                                                                                                                    |

**Notable disconnect:** the numbered scripts (`02b` → `03`/`04`) suggest preprocessing is a required step, but the actual trainers still load JSONL directly via `chunked_loader.py`. Only `train_decisions_ui.py` uses the mmap output — meaning **`preprocess_trajectories.py` runs but its output is barely consumed** in the production headless flow. That's a strong consolidation target.

---

## Recommended cleanup targets (ranked by leverage)

1. **`chunked_loader.py` vs `mmap_dataset.py` duplication** — pick mmap as canonical; delete the inline zone-parsing in `chunked_loader` and `train_decisions.py`; route every trainer through the mmap dataset.
2. **`train_decisions_ui.py` (2,730 lines)** — split GUI off from training; the training core should be importable from `train_decisions.py` and shared.
3. **Three recording controllers** (`RecordingPlayerController`, `RecordingControllerAi`, `RecordingLobbyPlayerAi`, plus the listener) — identify dead ones, delete; keep one.
4. **`preprocess_trajectories.py` (669 lines, zero tests, owns the binary format)** — add a round-trip test on 3 synthetic JSONLs that asserts the resulting mmap arrays decode back to the original records.
5. **Magic-number contracts** — centralize `GLOBAL_DIM=96`, `CARD_DIM=256`, `ACTION_DIM=64`, zone caps, `GAME_STATE_DIM=37216` in one Python module and import everywhere (currently redefined in ≥5 places). Add a Java equivalent that's read at start-up to detect drift.
6. **`SimulateRLTraining.java` (956 lines, 6 modes in one file)** — extract per-mode classes; the `collect` mode is the one supervised training depends on, and it's buried.
